rigging/dbus_connection.py

```diff
@@ -0,0 +1,105 @@
+import dbus
+import dbus.service
+import dbus.mainloop.glib
+from gi.repository import GLib
+from rigging.exceptions import (DBusServiceExistsError, 
+                            DBusServiceDoesntExistError,
+                            DBusMethodDoesntExistError)
+
+class RigDBusMessage:
+    result = None
+    success = None
+
+    def __init__(self, result=None, success=None):
+        self.result = result
+        self.success = success
+
+class RigDBusCommand:
+    name = None
+
+class RigDBusCommandDestroy(RigDBusCommand):
+    name = "destroy"
+
+class RigDBusConnection:
+    """
+    Used to abstract communication with an existing rig over the dbus
+    service created by that particular rig.
+    """
+    def __init__(self, rig_name):
+        """
+        :param rig_name: The name of the rig, which correlates to the name
+        of the socket
+        """
+        self.name = rig_name
+
+        # TODO: Check/handle errors
+        self._bus = dbus.SessionBus()
+        try:
+            self._rig = self._bus.get_object(
+                        f"com.redhat.Rig.{rig_name}", f"/RigControl")
+        except dbus.exceptions.DBusException as exc:
+            if exc.get_dbus_name() == "org.freedesktop.DBus.Error.ServiceUnknown":
+                raise DBusServiceDoesntExistError(rig_name)
+            raise exc
+
+
+
+    def _communicate(self, command):
+        """
+        Send a rig an instruction and then return the result directly to the
+        calling command for further handling
+
+        :param command: The command to have the rig perform
+        :return: The result of the command
+        """
+        method = getattr(self._rig, command.name)
+
+        try:
+            ret = method(dbus_interface="com.redhat.RigInterface")
+            return RigDBusMessage(**ret)
+
+        except dbus.exceptions.DBusException as exc:
+            if exc.get_dbus_name() == "org.freedesktop.DBus.Error.UnknownMethod":
+                raise DBusMethodDoesntExistError(f"{command.name}()")
+            raise exc
+
+        except Exception as exc:
+            print(f"Exception caught while calling {command} on {self.name}:{exc}")
+
+        return None
+
+    def destroy_rig(self):
+        """
+        Instruct the rig to self-terminate without triggering any configured
+        actions or generating an archive.
+        """
+
+        return self._communicate(RigDBusCommandDestroy())
+
+
+class RigDBusListener(dbus.service.Object):
```

Comment:
This is my hesitation at the moment - do we make the listener the dbus object, or the rig as a whole?

If we make only the listener the object, we do improve our self-protection against malicious messages from the bus, but we'd need to figure out a way to grant the listener access to the rig instance for most methods.

For instance, for `destroy` I'm thinking we'll use a poison pill here, but we'd need to give the listener a way to leverage that. Same thing for `trigger` to manually trigger a rig from the CLI. We'd also need to provide a way for the listener to get status information from the rig.

Do we just pass in specific methods to the listener when it is created? Or is there a cleaner way to manipulate the rig while keeping it separate from the dbus object?

Reply:
Done in 3309bdd6558064442d0ddeec5235936fdc1da49b and dd9bdd31ec5b9389f3a976939e47f221d89b889e

- `RigDBusListener` now has a dict  `_command_map`
- Callbacks are defined via `RigDBusListener.map_rig_command()` after the listener instance is created.
- Command `destroy` is mapped on `BaseRig` so it points to `BaseRig._destroy_self()`, which now just calls `BaseRig._exit(0)`